service_capacity_modeling/models/org/netflix/elasticsearch.py

```diff
@@ -0,0 +1,471 @@
+        instance=instance,
+        drive=drive,
+        needed_cores=int(requirement.cpu_cores.mid),
+        needed_disk_gib=int(requirement.disk_gib.mid),
+        needed_memory_gib=int(requirement.mem_gib.mid),
+        needed_network_mbps=requirement.network_mbps.mid,
+        # Assume that by provisioning enough memory we'll get
+        # a 90% hit rate, but take into account the reads per read
+        # from the per node dataset using leveled compaction
+        # FIXME: I feel like this can be improved
+        required_disk_ios=lambda x: _es_io_per_read(x) * math.ceil(0.1 * rps),
+        # Elasticsearch requires ephemeral disks to be % full because leveled
+        # compaction can make progress as long as there is some headroom
+        required_disk_space=lambda x: x * 1.25,
+        max_local_disk_gib=max_local_disk_gib,
+        # elasticsearch clusters can autobalance via shard placement
+        cluster_size=lambda x: x,
+        min_count=1,
+        # Sidecars/System takes away memory from elasticsearch
+        # Elasticsearch uses half of available system max of 32 for compressed
+        # oops
+        reserve_memory=lambda x: base_mem + max(32, x / 2),
+        core_reference_ghz=requirement.core_reference_ghz,
+    )
+
+    # Communicate to the actual provision that if we want reduced RF
+    params = {"elasticsearch.copies": copies_per_region}
+    _upsert_params(cluster, params)
+
+    # elasticsearch clusters generally should try to stay under some total number
+    # of nodes. Orgs do this for all kinds of reasons such as
+    #   * Security group limits. Since you must have < 500 rules if you're
+    #       ingressing public ips)
+    #   * Maintenance. If your restart script does one node at a time you want
+    #       smaller clusters so your restarts don't take months.
+    #   * NxN network issues. Sometimes smaller clusters of bigger nodes
+    #       are better for network propagation
+    if cluster.count > (max_regional_size // zones_per_region):
+        return None
+
+    ec2_cost = zones_per_region * cluster.annual_cost
+
+    cluster.cluster_type = "elasticsearch-data"
+    clusters = Clusters(
+        total_annual_cost=round(Decimal(ec2_cost), 2),
+        zonal=[cluster] * zones_per_region,
+        regional=list(),
+    )
+
+    return CapacityPlan(
+        requirements=Requirements(zonal=[requirement] * zones_per_region),
+        candidate_clusters=clusters,
+    )
+
+
+class NflxElasticsearchCapacityModel(CapacityModel):
+    @staticmethod
+    def capacity_plan(
+        instance: Instance,
+        drive: Drive,
+        context: RegionContext,
+        desires: CapacityDesires,
+        extra_model_arguments: Dict[str, Any],
+    ) -> Optional[CapacityPlan]:
+        # (FIXME): Need elasticsearch input
+        # TODO: Use durability requirements to compute RF.
+        copies_per_region: int = _target_rf(
+            desires, extra_model_arguments.get("copies_per_region", None)
```

Comment:
Existing ES self service assumes there will be 3 copies, for large use cases it makes sense to run with 2 copies to save on cluster cost. We can assume it's 3 by default and let users change it to 2 if they want.

Reply:
this also assumes 3 by default, will only reduce if they reduce their durability requirement.